src/runtime/wasm.ts

```diff
@@ -651,11 +668,42 @@ export class WasmParticle extends Particle {
     }
   }
 
+  // Wasm particles can request service calls with a Dictionary of arguments and an optional string
+  // tag to disambiguate different requests to the same service call.
+  async serviceRequest(callPtr: WasmAddress, argsPtr: WasmAddress, tagPtr: WasmAddress) {
+    const call = this.container.read(callPtr);
+    const args = new StringDecoder().decodeDictionary(this.container.read(argsPtr));
+    const tag = this.container.read(tagPtr);
```

Comment:
Can we generate a default value for the tag in case it isn't supplied?

Reply:
That's already done on the C++ side.

However, this code should be robust against incorrect usage (either accidental or malicious), so all of the wasm-to-js calls should have better parameters guards. I'll add a TODO.